llvm/docs/GitHub.rst

```diff
@@ -46,11 +73,106 @@ The default commit message for a squashed pull request is the pull request
 description, so this will allow reviewers to review the commit message before
 approving the commit.
 
+When pushing to your branch, make sure you push to the correct fork. Check your
+remotes with:
+
+::
+
+  git remote -v
+
+And make sure you push to the remote that's pointing to your fork.
+
+Rebasing Pull Requests and Force Pushes
+---------------------------------------
+In general, you should avoid rebasing a Pull Request and force pushing to the
+branch that's the root of the Pull Request during the review. This action will
+make the context of the old changes and comments harder to find and read.
+
+Sometimes, a rebase might be needed to update your branch with a fix for a test
+or in some dependent code.
+
+After your PR is reviewed and accepted, you want to rebase your branch to ensure
+you won't encounter merge conflicts when landing the PR.
+
+Landing your change
+-------------------
+When your PR has been accepted you can use the web interface to land your change.
+The button that should be used is called `Squash and merge` and after you can
```

Comment:
When using "Squash and merge", does it preserve the commit messages, or at least give a chance to edit the final commit message? This is quite vital. With the given procedures, I would expect the first commit in the branch to have the intended commit message, and later commits to be fixups. In such a case, we'd want to maintain the full commit message of the first commit, and skip the commit messages of the later fixup commits.

Dropping the commit message of the individual commits and summarizising it into a list of commits subjects like this, would be entirely unacceptable:

> Merge PR #1234
>
> Change the frobnicator
>
> Fixup
>
> Fixup

Reply:
Agreed, we probably need some guidance here on how to ensure a good commit message get's associated with the commit that ends up landing on main.

IIRC, github will offer an editable text box in the webui that contains the concatenation of all commit messages in the PR. Assuming the first commit contains a well-written commit message, and the follow-ons are simply fixups, it should be easy to simple remove the "fixup" commit messages.

Reply:
(As an aside: if it would take too long to research the details needed for writing proper documentation, I think it would be fine to write a placeholder TODO/FIXME in the documentation, e.g. stating that we need more documentation on how to ensure a good commit message, and land this PR, so that the documentation that is ready becomes available sooner rather than later).

Reply:
> I have no clue what it does when you have multiple commits in a branch and opening a PR, I've not done this before and we should probably not encourage anyone doing that.

I think it takes the title and the body from the first commit in that branch to use as a title and commit message.